overlord/patch/patch.go

```diff
@@ -40,45 +53,191 @@ func Init(s *state.State) {
 		panic("internal error: expected empty state, attempting to override patch-level without actual patching")
 	}
 	s.Set("patch-level", Level)
+
+	if s.Get("patch-sublevel", new(int)) != state.ErrNoState {
+		panic("internal error: expected empty state, attempting to override patch-sublevel without actual patching")
+	}
+	s.Set("patch-sublevel", Sublevel)
+}
+
+// applySublevelPatches applies all sublevel patches for given level, starting
+// from firstSublevel index.
+func applySublevelPatches(level, firstSublevel int, s *state.State) error {
+	for sublevel := firstSublevel; sublevel < len(patches[level]); sublevel++ {
+		if sublevel > 0 {
+			logger.Noticef("Patching system state level %d to sublevel %d...", level, sublevel)
+		}
+		err := applyOne(patches[level][sublevel], s, level, sublevel)
+		if err != nil {
+			logger.Noticef("Cannot patch: %v", err)
+			return fmt.Errorf("cannot patch system state to level %d, sublevel %d: %v", level, sublevel, err)
+		}
+	}
+	return nil
+}
+
+type patchSnapState struct {
+	Sequence []*patchSideInfo `json:"sequence"`
+	Current  snap.Revision    `json:"current"`
+}
+
+type patchSideInfo struct {
+	Revision snap.Revision `json:"revision"`
+}
+
+// lastIndex returns the last index of the given revision in the
+// snapst.Sequence
+func (snapst *patchSnapState) lastIndex(revision snap.Revision) int {
+	for i := len(snapst.Sequence) - 1; i >= 0; i-- {
+		if snapst.Sequence[i].Revision == revision {
+			return i
+		}
+	}
+	return -1
+}
+
+// maybeResetSublevelForLevel60 checks if the previously installed core revision was a patch 6 sublevel 0
+// i.e. unaware of sublevels and the need for resetting sublevel state and if so, resets sublevel to 0
+// to re-apply sublevel patches.
+func maybeResetSublevelForLevel60(s *state.State, sublevel *int) error {
+	s.Lock()
+	defer s.Unlock()
+
+	var snaps map[string]*json.RawMessage
+	err := s.Get("snaps", &snaps)
+	if err == state.ErrNoState {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	raw, ok := snaps["core"]
+	if err == state.ErrNoState || !ok {
```

Comment:
why the err check here? is checked just above

Reply:
Indeed, not needed, thanks.